src/librustc_mir/interpret/step.rs

```diff
@@ -11,8 +11,20 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
     pub fn inc_step_counter_and_check_limit(&mut self, n: usize) {
         self.terminators_remaining = self.terminators_remaining.saturating_sub(n);
         if self.terminators_remaining == 0 {
-            // FIXME(#49980): make this warning a lint
-            self.tcx.sess.span_warn(self.frame().span, "Constant evaluating a complex constant, this might take some time");
+            let node_id = self
+                    .stack()
+                    .iter()
+                    .rev()
+                    .filter_map(|frame| self.tcx.hir.as_local_node_id(frame.instance.def_id()))
+                    .next()
+                    .expect("some part of a failing const eval must be local");
+
+            self.tcx.struct_span_lint_node(
```

Comment:
So this alone won't cause const eval to abort. You should be able to run

```rust
tcx.lint_level_at_node(lint, id).0 == Level::Deny
```

to figure out whether the lint is denied.

This means you need to do that check for all stack frames (and not just the first local one), then report the lint at the most extreme level (forbid/deny) and if it is either forbid or deny, return a constant evaluation error here.

if all lint levels are allow or warn, continue the evaluation.

Reply:
N.B. There exists more than a single error lint level (e.g. `Level::Fatal` iirc).